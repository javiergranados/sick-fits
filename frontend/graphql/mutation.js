import gql from 'graphql-tag';

const CREATE_ITEM = gql`
  mutation createItem($title: String!, $description: String!, $price: Int!, $image: String, $largeImage: String) {
    createItem(title: $title, description: $description, price: $price, image: $image, largeImage: $largeImage) {
      id
    }
  }
`;

const UPDATE_ITEM = gql`
  mutation updateItem($id: ID!, $title: String, $description: String, $price: Int) {
    updateItem(id: $id, title: $title, description: $description, price: $price) {
      id
      title
      description
      price
    }
  }
`;

const DELETE_ITEM = gql`
  mutation deleteItem($id: ID!) {
    deleteItem(id: $id) {
      id
      title
    }
  }
`;

const SIGN_UP = gql`
  mutation signup($email: String!, $name: String!, $password: String!) {
    signup(email: $email, name: $name, password: $password) {
      id
      email
      name
    }
  }
`;

const SIGN_IN = gql`
  mutation signin($email: String!, $password: String!) {
    signin(email: $email, password: $password) {
      id
      email
      name
      password
    }
  }
`;

const SIGN_OUT = gql`
  mutation signout {
    signout {
      message
    }
  }
`;

const REQUEST_RESET = gql`
  mutation requestReset($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`;

export { CREATE_ITEM, UPDATE_ITEM, DELETE_ITEM, SIGN_UP, SIGN_IN, SIGN_OUT, REQUEST_RESET };
