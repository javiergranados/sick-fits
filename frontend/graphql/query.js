import gql from 'graphql-tag';
import { perPage } from '../config';

const GET_ITEMS = gql`
  query($skip: Int = 0, $first: Int = ${perPage}) {
    items(skip: $skip, first: $first, orderBy: createdAt_DESC) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

const GET_ITEM = gql`
  query($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      price
      largeImage
    }
  }
`;

const PAGINATION = gql`
  {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

const CURRENT_USER = gql`
  {
    me {
      id
      name
      email
      permissions
    }
  }
`;

const GET_ALL_USERS = gql`
  {
    users {
      id
      name
      email
      permissions
    }
  }
`;

export { GET_ITEMS, GET_ITEM, PAGINATION, CURRENT_USER, GET_ALL_USERS };
