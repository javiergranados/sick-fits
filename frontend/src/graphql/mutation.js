import gql from 'graphql-tag';

const CREATE_ITEM = gql`
  mutation createItem($title: String!, $description: String!, $price: Int!, $image: String, $largeImage: String) {
    createItem(title: $title, description: $description, price: $price, image: $image, largeImage: $largeImage) {
      id
    }
  }
`;

const UPDATE_ITEM = gql`
  mutation updateItem($id: ID!, $title: String, $description: String, $price: Int) {
    updateItem(id: $id, title: $title, description: $description, price: $price) {
      id
      title
      description
      price
    }
  }
`;

const DELETE_ITEM = gql`
  mutation deleteItem($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

export { CREATE_ITEM, UPDATE_ITEM, DELETE_ITEM };
