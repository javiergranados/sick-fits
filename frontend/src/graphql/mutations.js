import gql from 'graphql-tag';

const CREATE_ITEM = gql`
  mutation createItem($title: String!, $description: String!, $price: Int!, $image: String, $largeImage: String) {
    createItem(title: $title, description: $description, price: $price, image: $image, largeImage: $largeImage) {
      id
    }
  }
`;

export { CREATE_ITEM };
