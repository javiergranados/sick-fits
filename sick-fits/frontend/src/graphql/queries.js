import gql from 'graphql-tag';

const GET_ITEMS = gql`
  {
    items {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

export { GET_ITEMS };
