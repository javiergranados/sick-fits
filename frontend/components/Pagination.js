import { useQuery } from '@apollo/react-hooks';
import Head from 'next/head';
import { PAGINATION } from '../graphql/query';
import Error from './Error';
import { perPage } from '../config';
import * as S from './styles/Pagination';

const Pagination = ({ page }) => {
  const { loading, error, data } = useQuery(PAGINATION);

  if (loading) return <p>Loading...</p>;
  if (error) return <Error error={error} />;

  const { count } = data.itemsConnection.aggregate;
  const pages = Math.ceil(count / perPage);

  return (
    <S.Pagination>
      <Head>
        <title>
          Sick fits! | Page {page} of {pages}
        </title>
      </Head>
      <p>
        Page {page} of {pages}
      </p>
    </S.Pagination>
  );
};

export default Pagination;
