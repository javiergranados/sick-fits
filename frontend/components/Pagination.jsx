import { useQuery } from '@apollo/react-hooks';
import Head from 'next/head';
import Link from 'next/link';
import { PAGINATION } from '../graphql/query';
import Error from './Error';
import { perPage } from '../config';
import * as S from './styles/Pagination';

const Pagination = ({ page }) => {
  const { loading, error, data } = useQuery(PAGINATION);

  if (loading) return <p>Loading...</p>;
  if (error) return <Error error={error} />;

  const { count } = data.itemsConnection.aggregate;
  const pages = Math.ceil(count / perPage);

  return (
    <S.Pagination data-test="pagination">
      <Head>
        <title>
          Sick fits! | Page {page} of {pages}
        </title>
      </Head>
      <Link
        href={{
          pathname: 'items',
          query: { page: page - 1 },
        }}
      >
        <a className="prev" aria-disabled={page <= 1}>
          ⬅ Prev
        </a>
      </Link>
      <p>
        Page {page} of <span className="totalPages">{pages}</span>
      </p>
      <p>{count} Items Total</p>
      <Link
        href={{
          pathname: 'items',
          query: { page: page + 1 },
        }}
      >
        <a className="next" aria-disabled={page >= pages}>
          Next ➡
        </a>
      </Link>
    </S.Pagination>
  );
};

export default Pagination;
