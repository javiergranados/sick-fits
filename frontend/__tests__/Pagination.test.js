import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import { MockedProvider } from '@apollo/react-testing';
import { act } from 'react-dom/test-utils';
import Pagination from '../components/Pagination';
import { PAGINATION } from '../graphql/query';

Router.router = {
  push() {},
  prefetch() {},
};

function makeMocksFor(length) {
  return [
    {
      request: { query: PAGINATION },
      result: {
        data: {
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ];
}

describe('<Pagination />', () => {
  it('displays a loading message', () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(1)}>
        <Pagination page={1} />
      </MockedProvider>
    );

    expect(wrapper.text()).toContain('Loading...');
  });

  it('renders pagination for 18 items', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await act(wait);
    wrapper.update();

    expect(wrapper.find('.totalPages').text()).toEqual('5');
    const pagination = wrapper.find('div[data-test="pagination"]');
    expect(toJSON(pagination)).toMatchSnapshot();
  });

  it('disables prev button on first page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await act(wait);
    wrapper.update();

    expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(true);
    expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
  });
  it('disables next button on last page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={5} />
      </MockedProvider>
    );
    await act(wait);
    wrapper.update();

    expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
    expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(true);
  });
  it('enables all buttons on a middle page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={3} />
      </MockedProvider>
    );
    await act(wait);
    wrapper.update();

    expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
    expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
  });
});
