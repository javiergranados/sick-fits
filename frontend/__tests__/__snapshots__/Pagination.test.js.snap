// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders pagination for 18 items 1`] = `
<div
  className="Pagination-sc-174k1su-0 jfypyi"
  data-test="pagination"
>
  <Head>
    <_default
      headManager={Object {}}
      inAmpMode={false}
      reduceComponentsToState={[Function]}
    />
  </Head>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 0,
        },
      }
    }
  >
    <a
      aria-disabled={true}
      className="prev"
      href="/items?page=0"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      ⬅ Prev
    </a>
  </Link>
  <p>
    Page 
    1
     of 
    <span
      className="totalPages"
    >
      5
    </span>
  </p>
  <p>
    18
     Items Total
  </p>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 2,
        },
      }
    }
  >
    <a
      aria-disabled={false}
      className="next"
      href="/items?page=2"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Next ➡
    </a>
  </Link>
</div>
`;
